lib/galaxy/managers/credentials.py

```diff
@@ -0,0 +1,178 @@
+from typing import (
+    List,
+    Optional,
+    Tuple,
+    Union,
+)
+
+from sqlalchemy import select
+from sqlalchemy.orm import aliased
+
+from galaxy.exceptions import RequestParameterInvalidException
+from galaxy.model import (
+    CredentialsGroup,
+    Secret,
+    UserCredentials,
+    Variable,
+)
+from galaxy.model.base import transaction
+from galaxy.model.scoped_session import galaxy_scoped_session
+from galaxy.schema.credentials import SOURCE_TYPE
+from galaxy.schema.fields import DecodedDatabaseIdField
+
+CredentialsModelList = List[Union[UserCredentials, CredentialsGroup, Variable, Secret]]
+
+
+class CredentialsManager:
+    """Manager object shared by controllers for interacting with credentials."""
+
+    def __init__(self, session: galaxy_scoped_session) -> None:
+        self.session = session
+
+    def get_user_credentials(
+        self,
+        user_id: DecodedDatabaseIdField,
+        source_type: Optional[SOURCE_TYPE] = None,
+        source_id: Optional[str] = None,
+        user_credentials_id: Optional[DecodedDatabaseIdField] = None,
+        group_id: Optional[DecodedDatabaseIdField] = None,
+    ) -> List[Tuple[UserCredentials, CredentialsGroup]]:
+        user_cred_alias, group_alias = aliased(UserCredentials), aliased(CredentialsGroup)
+        stmt = (
+            select(user_cred_alias, group_alias)
+            .join(group_alias, group_alias.user_credentials_id == user_cred_alias.id)
+            .where(user_cred_alias.user_id == user_id)
+        )
+        if source_type:
+            stmt = stmt.where(user_cred_alias.source_type == source_type)
+        if source_id:
```

Comment:
I would suggest moving this to the very top of the method:
```py
if source_id and not source_type`:
    raise...
```
That way you exit early and don't mix argument validation logic with building a nontrivial db statement. Easier to read!